content/posts/fpcast.md

```diff
@@ -0,0 +1,429 @@
+dynamic call, it is determined by a variable.
+
+In web assembly, a function pointer is an index into a table of functions. There
+is a separate wasm instruction called `call_indirect` which takes an immediate
+(compile-time determined) argument to indicate what the signature of the
+function being called should be. The web assembly runtime checks that the
+function being called has a signature which matches the asserted signature and
+if it doesn't it throws an error.
+
+So we call `my_noargs_function2` through a `call_indirect` instruction which
+asserts that it takes two arguments. The wasm runtime checks and sees that in
+fact, `my_noargs_function2` only expects one argument, and it crashes with an
+"indirect call signature mismatch" error.
+
+## Ways to fix the call signature mismatch errors
+
+### 1. EMULATE_FUNCTION_POINTER_CASTS
+
+The Emscripten compiler comes with a flag that makes it emulate function pointer
+casting. This is what we have been using to resolve the troubles. However, it
+comes with a high cost: it is slow, it uses up large amounts of stack space, and
+it increases code size by a lot. Also, it makes calling a function pointer from
+Javascript inconvenient, and in particular breaks dynamic linking and many
+Emscripten ports which implement C libraries with Javascript code. I will
+describe later how this works and why it imposes such large costs.
+
+### 2. Just patch the packages
+
+We could just go through and patch the Python C packages to fix their function
+declarations. However, there is no easy way to detect these function pointer
+casts (there's no compiler setting that can generate warnings about them). It is
+also very tedious to locate the problematic code. The upstream packages are
+generally happy to accept these patches but it takes time to review them and
+they may not get backported as far as we need them (Pyodide is stuck on pretty
+old versions of some packages). So it generates a high maintentance burden. This
+solution isn't really workable for us.
+
+### 3. Automate the patching
+
+We can write code to analyze the `clang` abstract syntax tree, locate the bad
+pointer casts, and either automatically rewrite the source. The idea here is
+that we scan the AST for declarations that look like:
+
+```C
+PyMethodDef my_method_def = {
+    "my_method", // the name
+    (PyCFunction)my_method, // The C function pointer
+    METH_NOARGS, // the calling convention
+};
+```
+
+where `my_method` has a number of arguments different than 2. Then we locate the
+definition of `my_method`, say it looks like:
+
+```C
+static PyObject*
+my_method(PyObject* self){
+    // ...
+}
+```
+and we patch it to look like:
+```C
+static PyObject*
+my_method(PyObject* self, PyObject *_ignored){
+    // ...
+}
+```
+
+We also patch each spot where `my_method` is called with one argument. We fix a
+handful of other common problematic declarations as well. If problematic code
```

Comment:
`If problematic code when the source is confusing ` should probably be rephrased a bit. Like `If the problematic source code is confusing`